### DISCUSSION TITLE
Feature request: show tags next to note name in `:ZkNotes`

### DISCUSSION THREAD

**Discussion by mawkler** [zk-org/zk-nvim#144] [Ideas]
### Verify

- [X] I searched the existing discussions for similar ideas

### Share your idea or feature request

I would love the ability to list any note tags next to the title in when doing `:ZkNotes`, i.e. tags from the note's header. This would ease fuzzy finding over the notes based on the name _and_ the tags at once.

Old behaviour:

```
   ╭─────────────────────── Results ───────────────────────╮
   │  My First Note                                        │
   │  My Second Note                                       │
   │> My Third Note                                        │
   ╰───────────────────────────────────────────────────────╯
   ╭────────────────────── Zk Notes ───────────────────────╮
   │ >                                              13 / 13│
   ╰───────────────────────────────────────────────────────╯
```

Suggested new behaviour:

```
   ╭─────────────────────── Results ───────────────────────╮
   │  My First Note                           [rust, code] │
   │  My Second Note                                       │
   │> My Third Note                               [config] │
   ╰───────────────────────────────────────────────────────╯
   ╭────────────────────── Zk Notes ───────────────────────╮
   │ >                                              13 / 13│
   ╰───────────────────────────────────────────────────────╯
```

So now if I for example type "noterust" it would filter out the note "My First Note".

I've managed to list notes with their tags with the following command, as a proof of concept:

```sh
zk list --format "{{title}}{{#if tags}} [{{#each tags}}{{this}}{{#if @last}}{{else}}, {{/if}}{{/each}}]{{/if}}"
```

After skimming through the code it seems like zk-nvim generates a call the language server when doing `:ZkNotes`. Is there any way to pass the equivalent to the `--format` flag to the language server?

Also, it would be neat if the tags have a different color.

**Comment by tjex** [comment]
I don't think passing something like `--format` to the command would be the way to go. There also isn't any implementation for this out of the box (to pass `--format`). That is zk cli specific.

I think to implement this it would be a case of extending the `zk.edit` command to include the tags array in the returned title string, so that the tags become searchable along with the title. https://github.com/zk-org/zk-nvim/blob/dbf4eeab55b08856c9d6b6722dbff39630bb35eb/lua/zk.lua#L130

But, there would also need to be some handling as to precedence. Like, should note titles be considered first in search, and then tags? I would think so. 

It should also have a toggle, so users can decide to include tags or not in their `ZkNotes` command.

I personally wouldn't use this feature. If I really can't find something (like a reference note for an obscure research paper / long lost blog post I've forgotten the name of), I prefer to filter by tag first, then select filter for the title.

**Reply by mawkler** [reply]
> I prefer to filter by tag first, then select filter for the title.

But this only lets you search by one tag. What if you want to match over multiple tags? Moreover, I often find that I don't remember exactly what title and tag combination I've used for something I'm looking for, and fuzzy searching over both would be very powerful for discoverability

> Should note titles be considered first in search, and then tags? I would think so

I agree that the title should be prioritized higher first

> It should also have a toggle

I agree

**Reply by tjex** [reply]
I never need to search over multiple tags, because I'm very strict with what I use tags for. 
But I understand completely that it's common to rely strongly on tags for filtering. 

E.g, I also thought some time ago to extend the `ZkTags` command that lets you multi select tags via Telescope and populate a subsequent Telescope search on enter. But I'm using a lot of my bandwidth for zk at the moment keeping up with effectively all the issues / discussions / triaging and also completely rehauling our documentation site. The other maintainers are also not managing to devote much time at the moment. So feature requests are hard to get rolling.

Is this feature request something you'd be willing to contribute?

**Reply by mawkler** [reply]
Sure, I could give it a try when I find the time 🙂

**Reply by mawkler** [reply]
Is there a built-in function for getting the tags of one given note?

**Reply by tjex** [reply]
Not directly, but [here](https://github.com/zk-org/zk-nvim/blob/dbf4eeab55b08856c9d6b6722dbff39630bb35eb/lua/zk/pickers/telescope.lua#L102) is where the tags are pulled in as results. If you compare to `M.show_notes_picker()`, which pulls in titles as the results, perhaps a solution is to merge them together (titles + "..." + tags). 

Also, ideally any implementation should also be for the other pickers (fzf / select / minipick)

Perhaps the inclusion of tags in the titles row for `M.show_notes_picker` can be enabled by a toggle, e.g `:ZkNotes {tags = {true}}`, which will open up notes picker, with tags included in the row?

**Reply by mawkler** [reply]
That makes sense to me 🙂

**Reply by tjex** [reply]
@mawkler it occurred to me that shafftig is calling zk from zk-nvim and using its flags. So you could potentially just call out to the zk cli and have it return formatted text from stdout.

See [here line 104](https://github.com/zk-org/zk-nvim/pull/171/files#diff-b9ceb70c13112f25f6476ab5b31fd883e49c0b8d4b293b3815db32dd2e4c252a)

But please note, I'm currently not convinced I'd want to merge this. So the work may stay on your side for your own pleasure. Reason being is that increasingly so the maintenance work is being handled by myself only (which is fine), but I need to then be more conservative about how much extra complexity makes it into the code base, how fast and for what long term value.

This is also because I'm unfortunately not such a skilled developer and don't have much experience yet. So I'm trying to be strategic and keep zk and zk-nvim at the very least robust... As it already offers all the fundamental features for a very effective zettelkasten workflow.

I would be much more interested in a contribution to allow multi-selection in `:ZkTags`. So the user can return notes of more than one tag. This would also be a half way solution for you, so perhaps it's something you would consider.

**Reply by mawkler** [reply]
@tjex I totally understand your hesitancy towards adding and maintaning new features. However, if it's possible for zk-nvim to call out to `zk` like you describe, perhaps we could simply expose the `--format` flag to the users and let them pass an argument to it from their options passed to `setup()`? This would be a relatively small change while making the plugin more extensible for the user. What do you say?

I'm personally not interested in multi-selection with `:ZkTags` because what I want is to be able to fuzzy-find over both the title and tags simultaneously.
